### ISSUE TITLE
Does not work when model has an attribute the same name as autocomplete field

### ISSUE THREAD

**Issue by onebree** [bigtunacan/rails-jquery-autocomplete#34]
- Rails 4.1.1
- Ruby 2.1.2
- `rails-jquery-autocomplete (1.0.3)`
- `jquery-rails (3.1.0)`
- `jquery-ui-rails (5.0.5)`
- `bootstrap_form (2.1.1)`

Someone had the same issue, as seen in the forum post below:
https://railsforum.com/topic/1873-rails-4-autocomplete-gem/

A `Proposal` has the attribute `company_id`. Given there are many (100+) companies, I do not want a select box. Instead, using this gem would be most beneficial. _Similar to the forum post_, the autocomplete field also shares a name with an already-present attribute. A proposal also has a `company_name` (soon to be deprecated, but nevertheless currently in database).

I have followed the instructions, and have no errors on my end. However, the autocomplete functionality is not working. When I attempt to create a new proposal and enter an existing company name, nothing happens. Could the shared name of attributes be conflicting with each other?

```ruby
# config/routes.rb
resources :proposals do
  get :autocomplete_company_name, on: :collection
end

# company.rb
def autocomplete_display
  name
end

# proposals_controller.rb
autocomplete :company, :name, full: true, limit: 5, display_value: :autocomplete_display
```

```haml
// proposals/_form.html.haml
= bootstrap_nested_form_for @proposal, as: :proposal, url: url do |f|
  // original:
  // = f.collection_select :company_id, Company.all.to_a, :id, :name
  = f.text_field :company_name, as: :autocomplete, url: autocomplete_company_name_proposals_path, id_element: '#company_id'
```

We have another app which uses this gem (successfully), and I noticed a few things:
1. `application.js` includes `jquery2`
2. It runs Rails 4.2.0, the minimum requirement for `jquery2`
3. It only uses `autocomplete_field_tag` to search for an element, _not_ attaching the result to a model.

I am hesitant to upgrade the Rails version, but if necessary I will.

**Comment by mauriciopasquier** [comment]
Do you have the company id anywhere in the form? Something like this:

    f.hidden_field :company_id, id: 'company_id'

**Comment by onebree** [comment]
@mauriciopasquier no I do not. `company_id` is referenced no where else in the form, except for what was placed in the first comment. But like I said, the field `proposals.company_name` exists -- I cannot remove it until next release of this app. (Allows for easier rollback of data)

Note: the form does not have any mention of the `proposals.company_name` field, either.

**Comment by mauriciopasquier** [comment]
Sorry, I didn't understood you. You say that the autocomplete isn't completing anything then?
I'm not familiar with this `bootstrap_nested_form_for` but you could try using the [regular helper](https://github.com/bigtunacan/rails-jquery-autocomplete#getting-the-object-id).

**Comment by onebree** [comment]
I will this and get back to you. That gem basically applies bootstrap styling to form fields.

**Comment by onebree** [comment]
@mauriciopasquier that is correct. The autocomplete field was not completing anything. It acted like a regular text field.

Some updates:
  1. I updated the app to use Rails 4.2.5 -- no difference
  2. Still using `bootstrap nested form for`, I called a special helper forcing the view to use Rails Form Helpers instead. -- no difference
  3. I stripped down my form to just the following, and no difference:

    ```haml
    = form_for @proposal, as: :proposal, url: url do |f|
      = f.hidden_field :type, value: 'TlsProposal'
      = f.text_field :company_name, as: :autocomplete, url: autocomplete_company_name_proposals_path, id_element: '#company_id'
    ```

  4. I changed the form field to the following, as you suggested, but no luck:

    ```haml
    = f.autocomplete_field :company_name, autocomplete_company_name_proposals_path, :id_element => '#company_id'
    ```

Could you elaborate on the `:id_element`? Must this element be present in the form as a hidden field?

**Comment by mauriciopasquier** [comment]
About `:id_element`: you must have some sort of field to store the id of the object to associate with. But if the server isn't returning anything (did you check if that's the case with network monitor from your browser dev tools or something like that?) your issue is somewhere else.. to narrow the scope you could double check that 1) there are no js erros in console log, 2) the field is getting the search event attached, 3) the page sends the query, 4) the server returns a response

**Comment by onebree** [comment]
@mauriciopasquier to clarify, whatever field `:id_element` points to, it must be present in the form? As a hidden field, or text entry?

**Comment by mauriciopasquier** [comment]
Exactly, it's a field where this gem saves the selected item's id. Try with a `text_field` to see it. It's in the [readme](https://github.com/bigtunacan/rails-jquery-autocomplete#getting-the-object-id) :)

**Comment by onebree** [comment]
Okay, that part was was a little confusing to understand. I thought it was creating the field _and_ filling it in.

**Comment by onebree** [comment]
Update: I tried again with my original setup. In dev tools, I see that the autocomplete gem is present, but I am getting a 404 GET error when entering a name:

Request URL: `http://localhost:3000/proposals/autocomplete_company_name?term=Te`

Response:
```html
<header>
  <h1>
    ActiveRecord::RecordNotFound
      in ProposalsController#autocomplete_company_name
  </h1>
</header>

<div id="container">
  <h2>Couldn&#39;t find Proposal without an ID</h2>
```

I am surprised I did not mention this before. A proposal `belongs_to :company`. Given the error message, it is trying to search through the collection `:companies` given the proposal in the controller. However, since I am on `#new`, the proposal is only built, not created.

**Comment by onebree** [comment]
Thank you for all the help. I would not have thought to check web dev tools. By doing so, I was able to pinpoint the error.

- I used most the same code from my [initial comment][1]
- The response in the previous comment points to `proposals#set_proposal`. I needed to exclude `:autcomplete_company_name` from that `before_filter` method.
- The params correctly sent as `proposal[company_id]: '148'`
- My form now includes the following:

```haml
= f.hidden_field :company_id
= f.autocomplete_field :company_name, autocomplete_company_name_proposals_path, id_element: '#proposal_company_id'
```

[1]: https://github.com/bigtunacan/rails-jquery-autocomplete/issues/34#issue-119239356
